File: python/253. Meeting Rooms II.py
import heapq


# Definition for an interval.
class Interval(object):
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e


class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        intervals.sort(key=lambda x: (x.start, x.end))
        count = 0
        visit = [False] * len(intervals)
        for i in range(len(intervals)):
            if visit[i]:
                continue
            count += 1
            visit[i] = True
            end = intervals[i].end
            for j in range(i + 1, len(intervals)):
                if visit[j]:
                    continue
                if end <= intervals[j].start:
                    end = intervals[j].end
                    visit[j] = True
        return count

File: cpp/251-300/253. Meeting Rooms II.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
    	vector<int> starts, ends;
    	for(auto interval : intervals){
    		starts.push_back(interval.start);
    		ends.push_back(interval.end);
    	}
    	sort(starts.begin(), starts.end());
    	sort(ends.begin(), ends.end());
    	int empty = 0, rooms = 0, available = 0;
    	for(auto start : starts){
    		while(ends[empty] <= start){
    			++empty;
    			++available;
    		}
    		available? --available: ++rooms;
    	}
    	return rooms;
    }
};